core/service/src/components.rs

```diff
@@ -268,12 +280,257 @@ impl<C: Components> OffchainWorker<Self> for C where
+
+					network.service().put_value(hashed_public_key, payload.into_bytes());
+				}
+
+				// Query addresses of other validators.
+				// TODO: Should non-validators also do this? Probably not a good default.
+				match client.runtime_api().authorities(&id) {
+					Ok(authorities) => {
+						for authority in authorities.iter() {
+							println!("==== querying dht for authority: {}", authority.to_string());
+							// TODO: Remove unwrap.
+							let hashed_public_key = libp2p::multihash::encode(
+								libp2p::multihash::Hash::SHA2256,
+								authority.to_string().as_bytes(),
+							).expect("public key hashing not to fail");
+
+							network.service().get_value(&hashed_public_key.clone());
+						}
+					},
+					Err(e) => {
+						println!("==== Got no authorities, but an error: {:?}", e);
+					}
+				}
+			}
+
+			// We poll `finality_notification_stream`, but we only take the last event.
+			let mut last = None;
+			while let Ok(Async::Ready(Some(item))) = finality_notification_stream.poll() {
+				last = Some(item);
+			}
+			if let Some(notification) = last {
+				network.on_block_finalized(notification.hash, notification.header);
+			}
+
+			// Poll the RPC requests and answer them.
+			while let Ok(Async::Ready(Some(request))) = rpc_rx.poll() {
+				match request {
+					rpc::apis::system::Request::Health(sender) => {
+						let _ = sender.send(rpc::apis::system::Health {
+							peers: network.peers_debug_info().len(),
+							is_syncing: network.service().is_major_syncing(),
+							should_have_peers,
+						});
+					},
+					rpc::apis::system::Request::Peers(sender) => {
+						let _ = sender.send(network.peers_debug_info().into_iter().map(|(peer_id, p)|
+							rpc::apis::system::PeerInfo {
+								peer_id: peer_id.to_base58(),
+								roles: format!("{:?}", p.roles),
+								protocol_version: p.protocol_version,
+								best_hash: p.best_hash,
+								best_number: p.best_number,
+							}
+						).collect());
+					}
+					rpc::apis::system::Request::NetworkState(sender) => {
+						let _ = sender.send(network.network_state());
+					}
+				};
+			}
+
+			// Interval report for the external API.
+			while let Ok(Async::Ready(_)) = status_interval.poll() {
+				let status = NetworkStatus {
+					sync_state: network.sync_state(),
+					best_seen_block: network.best_seen_block(),
+					num_sync_peers: network.num_sync_peers(),
+					num_connected_peers: network.num_connected_peers(),
+					num_active_peers: network.num_active_peers(),
+					average_download_per_sec: network.average_download_per_sec(),
+					average_upload_per_sec: network.average_upload_per_sec(),
+				};
+				let state = network.network_state();
+
+				status_sinks.lock().retain(|sink| sink.unbounded_send((status.clone(), state.clone())).is_ok());
+			}
+
+			let authorities = client.runtime_api().authorities(&BlockId::hash(client.info().chain.best_hash));
```

Comment:
For GRANDPA we need to check the active authority set from the node state and not the runtime, it is tracked [here](https://github.com/paritytech/substrate/blob/master/core/finality-grandpa/src/environment.rs#L375). Maybe we can just pass a `Fn() -> Vec<Authority>` into this function and then build up the authorities externally (babe authorities + grandpa authorities).